src/drivers/intel/dmic.c

```diff
@@ -579,9 +578,22 @@ static int select_mode(struct dai *dai,
 	else
 		mfir = modes->mfir_b;
 
-	mmin = find_min_int16(mfir, modes->num_of_modes);
-	count = find_equal_int16(idx, mfir, mmin, modes->num_of_modes, 0);
-	n = idx[count - 1];
+	/* Search fir_list[] decimation factors from start towards end. The found
+	 * last configuration entry with searched decimation factor will be used.
+	 */
+	count = 0;
+	for (n = 0; fir_list[n]; n++) {
+		count = find_equal_int16(idx, mfir, fir_list[n]->decim_factor,
+					 modes->num_of_modes, 0);
+		if (count > 0)
+			break;
+	}
+
+	if (!count) {
+		dai_err(dai, "select_mode(): No filter for decimate by %d found", mfir);
+		return -EINVAL;
+	}
+	n = idx[count - 1]; /* Option with highest clock divisor and lowest mic clock rate */
```

Comment:
and this is not needed since ```n``` should already have the correct value ?

Reply:
They are not the same, with this change:

```
-       n = idx[count - 1]; /* Option with highest clock divisor and lowest mic clock rate */
+       dai_err(dai, "select_mode(): n = %d", n);
+       n = idx[found - 1]; /* Option with highest clock divisor and lowest mic clock rate */
+       dai_err(dai, "select_mode(): n = %d", n);

```

The trace output is:

```
..
[  4367496.041667] (       29.916666) c0 dmic-dai     2.0   src/drivers/intel/dmic.c:595  ERROR select_mode(): n = 0
[  4367514.583333] (       18.541666) c0 dmic-dai     2.0   src/drivers/intel/dmic.c:597  ERROR select_mode(): n = 8
..
[  4368280.250000] (       48.416668) c0 dmic-dai     2.1   src/drivers/intel/dmic.c:595  ERROR select_mode(): n = 0
[  4368299.375000] (       19.125000) c0 dmic-dai     2.1   src/drivers/intel/dmic.c:597  ERROR select_mode(): n = 3
..
```